### ISSUE TITLE
GLPI Agent 1.4 tls_process_server_certificate:certificate verify failed - using properly chained RapidSSL cert

### ISSUE THREAD

**Issue by pmatis** [glpi-project/glpi-agent#226]
### Documentation

Yes, I read it

### Solution search

Yes, I read it

### Professional support

Yes, I know

### Is your help request related to a problem?
```
[Wed Oct  5 13:24:21 2022][info] sending prolog request to server0
[Wed Oct  5 13:24:21 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>DESKTOP-G4KLOIS-2022-10-05-13-07-32</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Wed Oct  5 13:24:22 2022][debug] [http client] Updating keystore known certificates
[Wed Oct  5 13:24:22 2022][debug2] Changing to 'C:/Program Files/GLPI-Agent/var/keystore-export-UtvXim' temporary folder
[Wed Oct  5 13:24:22 2022][debug2] executing certutil -Store -Silent -Split
[Wed Oct  5 13:24:22 2022][debug2] executing certutil -Store -Silent -Enterprise -Split
[Wed Oct  5 13:24:22 2022][debug2] executing certutil -encode 109f1caed645bb78b3ea2b94c0697c740733031c.crt temp.cer
[Wed Oct  5 13:24:22 2022][debug2] executing certutil -encode d559a586669b08f46a30a133f8a9ed3d038e2ea8.crt temp.cer
[Wed Oct  5 13:24:22 2022][debug2] executing certutil -encode fee449ee0e3965a5246f000e87fde2a065fd89d4.crt temp.cer
[Wed Oct  5 13:24:23 2022][debug2] Changing back to 'C:/Program Files/GLPI-Agent/perl/bin' folder
[Wed Oct  5 13:24:23 2022][error] [http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor), SSL connect attempt failed error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
```

### Expected behavior

Verify the cert successfully and connect

### Actions you've considered

Testing my certificate chain here: `https://www.rapidsslonline.com/ssl-tools/ssl-checker.php`
It passed their tests so the certificate is fully installed and properly chained to a known CA.

### Additional context

Edge, Firefox, Chrome and Opera all report a proper certificate. There must be something I can do to get this working, without using no-ssl-check, or other registry entries that require maintenance every time the cert is renewed.

If someone willing to help wants the actual URL being used, just ask. I won't publish it in this request, however.

I know there must be a simple solution, but I can't see what it is right now. Thank you!

**Comment by g-bougard** [comment]
Hi @pmatis 
this is strange: I generally expect more debug on SSL connection when debug2 is enabled.
The `Updating keystore known certificates` is new to 1.3 version. Can you confirm which of 1.3, 1.4 or nightly built version you're using ?
Indeed, I suspect you're still using 1.3 version as 1.4 fixes an issue with the keystore support. If this is the case, just upgrade your agent to 1.4.
If you're still using 1.4, can you try the 1.2 version to check if there's a bug with new SSL features in your case ?

**Comment by g-bougard** [comment]
I see now `GLPI Agent 1.4` in the title. So try the 1.2 to answer my last question.

**Comment by sbsroc** [comment]
Hmmm, I tried 1.2 but there's a similar error. I received "internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)" with 1.4 as well, then after a while (hours, perhaps?) I saw the certificate messages. I

```
[Thu Oct  6 11:45:23 2022][info] target server0: server https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:23 2022][debug] [http client] Using Compress::Zlib for compression
[Thu Oct  6 11:45:23 2022][info] sending prolog request to server0
[Thu Oct  6 11:45:23 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Thu Oct  6 11:45:24 2022][error] [http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)
[Thu Oct  6 11:45:24 2022][error] No answer from server at https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:24 2022][info] target server0: next run: Thu Oct  6 11:46:24 2022 - https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:24 2022][debug] [http server] GET request / from client 127.0.0.1
[Thu Oct  6 11:45:24 2022][debug] [http server] response status 200
```

When I set no-ssl-check to 1 in the registry and restart version 1.2, I see this from a manual inventory:

```
[Thu Oct  6 11:49:44 2022][info] sending prolog request to server0
[Thu Oct  6 11:49:44 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Thu Oct  6 11:49:44 2022][debug] [http client] authentication required, submitting credentials
[Thu Oct  6 11:49:44 2022][debug] [http client] authentication required, trying basic with xxxxxxxxxxxxx user (MSP Portal)
[Thu Oct  6 11:49:44 2022][debug2] format: JSON
[Thu Oct  6 11:49:44 2022][debug2] [http client] receiving message:
{"status":"ok","expiration":18}
[Thu Oct  6 11:49:45 2022][info] server0 answer shows it supports GLPI Agent protocol
[Thu Oct  6 11:49:45 2022][debug] new thread 3 to handle task Inventory
[Thu Oct  6 11:49:45 2022][info] running task Inventory
[Thu Oct  6 11:49:45 2022][debug] [http server] GET request / from client 127.0.0.1
[Thu Oct  6 11:49:45 2022][info] New inventory from Presentation.theirdomain.local-2022-10-04-20-36-41 for server0
[Thu Oct  6 11:49:45 2022][debug] [http server] response status 200
[Thu Oct  6 11:49:45 2022][debug2] module GLPI::Agent::Task::Inventory::AIX disabled
[Thu Oct  6 11:49:45 2022][debug2]   GLPI::Agent::Task::Inventory::AIX::Bios disabled ...
```
So, even though it's giving a "Bad file descriptor" error, it seems to be a cert issue there, too.

**Comment by pmatis** [comment]
Hmmm, I tried 1.2 but there's a similar error. I received "internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)" with 1.4 as well, then after a while (hours, perhaps?) I saw the certificate messages. It seems to be the same pattern, but not sure what triggers the change.

```
[Thu Oct  6 11:45:23 2022][info] target server0: server https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:23 2022][debug] [http client] Using Compress::Zlib for compression
[Thu Oct  6 11:45:23 2022][info] sending prolog request to server0
[Thu Oct  6 11:45:23 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Thu Oct  6 11:45:24 2022][error] [http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)
[Thu Oct  6 11:45:24 2022][error] No answer from server at https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:24 2022][info] target server0: next run: Thu Oct  6 11:46:24 2022 - https://portal.mydomain.com/front/inventory.php
[Thu Oct  6 11:45:24 2022][debug] [http server] GET request / from client 127.0.0.1
[Thu Oct  6 11:45:24 2022][debug] [http server] response status 200
```

When I set no-ssl-check to 1 in the registry and restart version 1.2, I see this from a manual inventory:

```
[Thu Oct  6 11:49:44 2022][info] sending prolog request to server0
[Thu Oct  6 11:49:44 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Thu Oct  6 11:49:44 2022][debug] [http client] authentication required, submitting credentials
[Thu Oct  6 11:49:44 2022][debug] [http client] authentication required, trying basic with xxxxxxxxxxxxx user (MSP Portal)
[Thu Oct  6 11:49:44 2022][debug2] format: JSON
[Thu Oct  6 11:49:44 2022][debug2] [http client] receiving message:
{"status":"ok","expiration":18}
[Thu Oct  6 11:49:45 2022][info] server0 answer shows it supports GLPI Agent protocol
[Thu Oct  6 11:49:45 2022][debug] new thread 3 to handle task Inventory
[Thu Oct  6 11:49:45 2022][info] running task Inventory
[Thu Oct  6 11:49:45 2022][debug] [http server] GET request / from client 127.0.0.1
[Thu Oct  6 11:49:45 2022][info] New inventory from Presentation.theirdomain.local-2022-10-04-20-36-41 for server0
[Thu Oct  6 11:49:45 2022][debug] [http server] response status 200
[Thu Oct  6 11:49:45 2022][debug2] module GLPI::Agent::Task::Inventory::AIX disabled
[Thu Oct  6 11:49:45 2022][debug2]   GLPI::Agent::Task::Inventory::AIX::Bios disabled ...
```
So, even though it's giving a "Bad file descriptor" error, it seems to be a cert issue there, too.

**Comment by g-bougard** [comment]
Are you sure the domain used as `server` option is the one your server expect ? I mean the same as the one set on your GLPI server certificate.

**Comment by pmatis** [comment]
Thank you for your reply. I was absolutely sure that the certificate was correct, since https://www.rapidsslonline.com/ssl-tools/ssl-checker.php tested my server and showed green between the cert, the intermediate and the CA. All good it said... but openssl on the CLI showed that I had a RapidSSL cert chained with a Let's Encrypt intermediate cert! All the web browsers accepted it! Anyway, that's fixed, but I'm still having the same error on the agent.

I'm not using GLPI server as I wrote my own scaled down app to collect the agent data. I reviewed the protocol docs and reverse engineered it a bit. All I need works without SSL turned on, or no_ssl_check set to 1. Still, there MUST be something I'm not doing right, but I can't imagine what that is!

I'd like to share my actual URL with you - perhaps you can spot something stupid I'm doing - you can test an agent with temp credentials I can share with you. Is there a place I can send it to you without it being logged in this request?

```
[Fri Oct  7 11:01:24 2022][debug] [target server0] New storage maintenance event on deploy task
[Fri Oct  7 11:01:24 2022][debug2] [target server0] Event scheduled in 120 seconds
[Fri Oct  7 11:01:25 2022][debug] target server0: init event for Collect task
[Fri Oct  7 11:01:39 2022][debug] [http server] GET request /now from client 127.0.0.1
[Fri Oct  7 11:01:39 2022][debug] [http server] rescheduling next contact for all targets right now
[Fri Oct  7 11:01:39 2022][debug] [http server] response status 200
[Fri Oct  7 11:01:40 2022][info] target server0: server https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:01:40 2022][debug] [http client] Using Compress::Zlib for compression
[Fri Oct  7 11:01:40 2022][info] sending prolog request to server0
[Fri Oct  7 11:01:40 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Fri Oct  7 11:01:40 2022][error] [http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)
[Fri Oct  7 11:01:40 2022][error] No answer from server at https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:01:40 2022][info] target server0: next run: Fri Oct  7 11:02:40 2022 - https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:01:40 2022][debug] [http server] GET request / from client 127.0.0.1
[Fri Oct  7 11:01:40 2022][debug] [http server] response status 200
[Fri Oct  7 11:02:40 2022][info] target server0: server https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:02:40 2022][debug] [http client] Using Compress::Zlib for compression
[Fri Oct  7 11:02:40 2022][info] sending prolog request to server0
[Fri Oct  7 11:02:40 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Fri Oct  7 11:02:40 2022][error] [http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)
[Fri Oct  7 11:02:40 2022][error] No answer from server at https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:02:40 2022][info] target server0: next run: Fri Oct  7 11:04:40 2022 - https://portal.mydomain.com/front/inventory.php
[Fri Oct  7 11:02:40 2022][debug] Agent memory usage before freeing memory: WSS=43503616 PFU=115118080
[Fri Oct  7 11:02:40 2022][info] GLPI Agent memory usage: WSS=3321856 PFU=115118080
```

**Comment by g-bougard** [comment]
Hi @pmatis 
you can share it on gbougard <at> teclib <dot> com. Will check it quickly but you're really in a very exotic context ;-)

**Comment by pmatis** [comment]
Thank you! I sent a reply. If it's the agent, I hope this helps to fix the bug. If it's something I'm doing incorrectly, hopefully we can gather enough information for me to dig in to fix it.
Merci beacoup!

**Comment by g-bougard** [comment]
Hi @pmatis,

from a linux station, just trying to contact your server to just send a PROLOG or contact request, I'm getting another error:
```
[error] [http client] communication error: 405 Method Not Allowed
```
Modifying the agent to directly send a CONTACT request I'm getting an error with a full page content:
```
[error] [http client] communication error: 405 Method Not Allowed
[debug2] [http client] received message:
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8" />
    <meta name="robots" content="noindex,nofollow,noarchive" />
    <title>An Error Occurred: Method Not Allowed</title>
    <style>body { background-color: #fff; color: #222; font: 16px/1.5 -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif; margin: 0; }
.container { margin: 30px; max-width: 600px; }
h1 { color: #dc3545; font-size: 24px; }
h2 { font-size: 18px; }</style>
</head>
<body>
<div class="container">
    <h1>Oops! An Error Occurred</h1>
    <h2>The server returned a "405 Method Not Allowed".</h2>

    <p>
        Something is broken. Please let us know what you were doing when this error occurred.
        We will fix it as soon as possible. Sorry for any inconvenience caused.
    </p>
</div>
</body>
</html>
[error] [http client] unexpected content, starting with: <!DOCTYPE html>
<html>
[error] No answer from server at https://....
```
And this means there's no problem with SSL there.
Hope this helps.

**Comment by pmatis** [comment]
Thank you for giving your time to look at this. I can't understand why you're receiving a 405 error. Did the agent contact /front/inventory.php? That's the only URL that wrote to respond appropriately. Did you run this as a service or from the command line? If it was a shell command, would you share it so I can reproduce the error as you experienced?

I worked on this for a while, trying to see if I could get wireshark to decode the SSL traffic so I could see what the client and server were doing, but I just couldn't figure that out. I'll have to try with a Linux client, but with a Windows client I'm brought back to the same thing. With no-ssl-check set to 0, I see:

`[http client] internal response: 500 Can't connect to portal.mydomain.com:443 (Bad file descriptor)`

If I set no-ssl-check to 1 and restart the service, the agent connects, runs all the actions and reports inventory back to the server as expected. While it might not be SSL related, perhaps there's a connection error that's only noticed when SSL cert checking is turned on.

**Comment by pmatis** [comment]
I decided to check things against a known good website. I chose google.com and found this:

```
[Sat Oct  8 23:40:11 2022][debug] [http client] Using Compress::Zlib for compression
[Sat Oct  8 23:40:11 2022][info] sending prolog request to server0
[Sat Oct  8 23:40:11 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Sat Oct  8 23:40:11 2022][error] [http client] internal response: 500 Can't connect to www.google.com:443 (Bad file descriptor)
[Sat Oct  8 23:40:11 2022][error] No answer from server at https://www.google.com/front/inventory.php
```

I then set no-ssl-log to 1 and got a proper 404 error:

```
[Sat Oct  8 23:41:12 2022][debug] [http client] Using Compress::Zlib for compression
[Sat Oct  8 23:41:12 2022][info] sending prolog request to server0
[Sat Oct  8 23:41:12 2022][debug2] [http client] sending message:
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <DEVICEID>Presentation.theirdomain.local-2022-10-04-20-36-41</DEVICEID>
  <QUERY>PROLOG</QUERY>
  <TOKEN>12345678</TOKEN>
</REQUEST>
[Sat Oct  8 23:41:12 2022][error] [http client] communication error: 404 Not Found
[Sat Oct  8 23:41:12 2022][error] No answer from server at https://www.google.com/front/inventory.php
```

I then looked at my server logs. When no-ssl-check was 0, no data was logged in acces.log or error.log. It seems that it's either not reaching out to the server at all, or failing on SSL/TLS negotiation. Am I missing something?

There seems to be something wrong with my Windows VM or the agent or dependencies, etc.

**Comment by g-bougard** [comment]
Hi @pmatis 

I didn't use /front/inventory.php on the url as this is not a requirement as explained here: https://glpi-agent.readthedocs.io/en/latest/configuration.html#server

Anyway, to test our openssl stack, can you try to run the following one liner from the agent installation folder in a cmd console ?
And share the resulting output ?
```
perl\bin\glpi-agent.exe -Iperl\agent -MConfig -MGLPI::Agent::Logger -MGLPI::Agent::HTTP::Client -MURI -MHTTP::Request -e "$url = shift @ARGV; %conf = @ARGV; $req = GLPI::Agent::HTTP::Client->new(logger => GLPI::Agent::Logger->new(%conf), config => \%conf); $test = $req->request(HTTP::Request->new(GET => URI->new($url))); undef $req; print 'STATUS: ', $test->status_line, \"\nperl uname: \" , $Config{myuname}" https://github.com/ debug 2 no-ssl-check 0
```
Indeed, `perl\bin\glpi-agent.exe` is the glpi-agent embedded `perl.exe` which has been renamed. The first argument at the end is the url (here `https://githhub.com` to test and the following arguments can define agent config to use. So you can also test on your own server.

It should return something like this:
```
C:\Program Files\GLPI-Agent>perl\bin\glpi-agent.exe -Iperl\agent -MConfig -MGLPI::Agent::Logger -MGLPI::Agent::HTTP::Client -MURI -MHTTP::Request -e "$url = shift @ARGV; %conf = @ARGV; $req = GLPI::Agent::HTTP::Client->new(logger => GLPI::Agent::Logger->new(%conf), config => \%conf); $test = $req->request(HTTP::Request->new(GET => URI->new($url))); undef $req; print 'STATUS: ', $test->status_line, \"\nperl uname: \" , $Config{myuname}" https://github.com/ debug 2 no-ssl-check 0
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.5-git23776baf)
[debug] [http client] Updating keystore known certificates
[debug2] Changing to 'C:/Users/gbougard/AppData/Local/Temp/keystore-export-pHD_Jr' temporary folder
[debug2] executing certutil -Store -Silent -Split
[debug2] executing certutil -Store -Silent -Enterprise -Split
[debug2] executing certutil -encode 109...1c.crt temp.cer
[debug2] executing certutil -encode 187...9b.crt temp.cer
[debug2] Changing back to 'C:/Program Files/GLPI-Agent' folder
DEBUG: .../IO/Socket/SSL.pm:2968: new ctx 56942880
DEBUG: .../IO/Socket/SSL.pm:704: socket not yet connected
DEBUG: .../IO/Socket/SSL.pm:706: socket connected
DEBUG: .../IO/Socket/SSL.pm:729: ssl handshake not started
DEBUG: .../IO/Socket/SSL.pm:771: using SNI with hostname github.com
DEBUG: .../IO/Socket/SSL.pm:806: request OCSP stapling
DEBUG: .../IO/Socket/SSL.pm:822: set socket to non-blocking to enforce timeout=180
DEBUG: .../IO/Socket/SSL.pm:836: call Net::SSLeay::connect
DEBUG: .../IO/Socket/SSL.pm:839: done Net::SSLeay::connect -> -1
DEBUG: .../IO/Socket/SSL.pm:849: ssl handshake in progress
DEBUG: .../IO/Socket/SSL.pm:859: waiting for fd to become ready: SSL wants a read first
DEBUG: .../IO/Socket/SSL.pm:879: socket ready, retrying connect
DEBUG: .../IO/Socket/SSL.pm:836: call Net::SSLeay::connect
DEBUG: .../IO/Socket/SSL.pm:2812: ok=1 [2] /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert Global Root CA/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert Global Root CA
DEBUG: .../IO/Socket/SSL.pm:2812: ok=1 [1] /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert Global Root CA/C=US/O=DigiCert Inc/CN=DigiCert TLS Hybrid ECC SHA384 2020 CA1
DEBUG: .../IO/Socket/SSL.pm:2812: ok=1 [0] /C=US/O=DigiCert Inc/CN=DigiCert TLS Hybrid ECC SHA384 2020 CA1/C=US/ST=California/L=San Francisco/O=GitHub, Inc./CN=github.com
DEBUG: .../IO/Socket/SSL.pm:1786: scheme=www cert=67558448
DEBUG: .../IO/Socket/SSL.pm:1796: identity=github.com cn=github.com alt=2 github.com 2 www.github.com
DEBUG: .../IO/Socket/SSL.pm:2859: did not get stapled OCSP response
DEBUG: .../IO/Socket/SSL.pm:839: done Net::SSLeay::connect -> 1
DEBUG: .../IO/Socket/SSL.pm:894: ssl handshake done
DEBUG: .../IO/Socket/SSL.pm:3017: free ctx 56942880 open=56942880
DEBUG: .../IO/Socket/SSL.pm:3021: free ctx 56942880 callback
DEBUG: .../IO/Socket/SSL.pm:3028: OK free ctx 56942880
STATUS: 200 OK
perl uname: Win32 strawberry-perl 5.36.0.1 #1 Wed Oct  5 13:54:23 2022 x64
C:\Program Files\GLPI-Agent>
```
I used here a development release but you should have the same kind of output.

**Comment by pmatis** [comment]
Excellent! I'll look in to the documentation you gave me on the URL, but I noticed that without anything trailing on the hostname, the agent auto-added /front/inventory.php. Maybe I can detect GLPI in the user-agent and redirect? I'll read and decide.

Anyway, I ran your command exactly and I have an error. We aren't getting the same thing, and at this point I've ruled out my server as the cause. That said, I'm not sure what the cause really is.

```
C:\Program Files\GLPI-Agent>perl\bin\glpi-agent.exe -Iperl\agent -MConfig -MGLPI::Agent::Logger -MGLPI::Agent::HTTP::Client -MURI -MHTTP::Request -e "$url = shift @ARGV; %conf = @ARGV; $req = GLPI::Agent::HTTP::Client->new(logger => GLPI::Agent::Logger->new(%conf), config => \%conf); $test = $req->request(HTTP::Request->new(GET => URI->new($url))); undef $req; print 'STATUS: ', $test->status_line, \"\nperl uname: \" , $Config{myuname}" https://github.com/ debug 2 no-ssl-check 0
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.4)
[debug] [http client] Updating keystore known certificates
[debug2] Changing to 'C:/Users/sswitzer/AppData/Local/Temp/keystore-export-GBZ3sM' temporary folder
[debug2] executing certutil -Store -Silent -Split
[debug2] executing certutil -Store -Silent -Enterprise -Split
[debug2] executing certutil -encode 109f1caed645bb78b3ea2b94c0697c740733031c.crt temp.cer
[debug2] executing certutil -encode d559a586669b08f46a30a133f8a9ed3d038e2ea8.crt temp.cer
[debug2] executing certutil -encode fee449ee0e3965a5246f000e87fde2a065fd89d4.crt temp.cer
[debug2] Changing back to 'C:/Program Files/GLPI-Agent' folder
DEBUG: .../IO/Socket/SSL.pm:2961: new ctx 49797792
DEBUG: .../IO/Socket/SSL.pm:704: socket not yet connected
DEBUG: .../IO/Socket/SSL.pm:706: socket connected
DEBUG: .../IO/Socket/SSL.pm:729: ssl handshake not started
DEBUG: .../IO/Socket/SSL.pm:771: using SNI with hostname github.com
DEBUG: .../IO/Socket/SSL.pm:806: request OCSP stapling
DEBUG: .../IO/Socket/SSL.pm:822: set socket to non-blocking to enforce timeout=180
DEBUG: .../IO/Socket/SSL.pm:836: call Net::SSLeay::connect
DEBUG: .../IO/Socket/SSL.pm:839: done Net::SSLeay::connect -> -1
DEBUG: .../IO/Socket/SSL.pm:849: ssl handshake in progress
DEBUG: .../IO/Socket/SSL.pm:859: waiting for fd to become ready: SSL wants a read first
DEBUG: .../IO/Socket/SSL.pm:879: socket ready, retrying connect
DEBUG: .../IO/Socket/SSL.pm:836: call Net::SSLeay::connect
DEBUG: .../IO/Socket/SSL.pm:2805: ok=0 [0] /CN=ESET SSL Filter CA/O=ESET, spol. s r. o./C=SK/C=US/ST=California/L=San Francisco/O=GitHub, Inc./CN=github.com
DEBUG: .../IO/Socket/SSL.pm:839: done Net::SSLeay::connect -> -1
DEBUG: .../IO/Socket/SSL.pm:842: SSL connect attempt failed

DEBUG: .../IO/Socket/SSL.pm:842: local error: SSL connect attempt failed error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
DEBUG: .../IO/Socket/SSL.pm:845: fatal SSL error: SSL connect attempt failed error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
DEBUG: .../lib/Net/HTTPS.pm:67: ignoring less severe local error 'IO::Socket::IP configuration failed', keep 'SSL connect attempt failed error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed'
DEBUG: .../IO/Socket/SSL.pm:3010: free ctx 49797792 open=49797792
DEBUG: .../IO/Socket/SSL.pm:3014: free ctx 49797792 callback
DEBUG: .../IO/Socket/SSL.pm:3021: OK free ctx 49797792
[error] [http client] internal response: 500 Can't connect to github.com:443 (Bad file descriptor), SSL connect attempt failed error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
STATUS: 500 Can't connect to github.com:443 (Bad file descriptor)
perl uname: Win32 strawberry-perl 5.36.0.1 #1 Wed Jun 15 15:36:57 2022 x64
C:\Program Files\GLPI-Agent>
```

I see a "certificate verify failed" message and "IO::Socket::IP configuration failed." It says that the latter is less severe, but perhaps it's setting up the first for failure? Likely not, as it's able to get the certificate information.

Could this be a US/English based Windows installation vs another?

**Comment by pmatis** [comment]
It's finally making sense and I think I've reached a point where I can pursue this on my own. ESET antivirus is installed on each of the machines I used and it seems to be hijacking the connection to examine the contents of the communications as evidenced by the certificate information that it sees. I'll see what I can do to turn off this feature. I think advanced anti-virus is the entirety of this issue.
Thank you for the command I used to uncover the root cause of this issue!
